File: Daily/725.split-linked-list-in-parts.js
/*
 * @lc app=leetcode id=725 lang=javascript
 *
 * [725] Split Linked List in Parts
 */
function ListNode(val, next) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
}
// @lc code=start
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {(ListNode|null)[]}
 */
var splitListToParts = function (head, k) {
    let listLength = 0;
    let node = head;

    while (node !== null) {
        listLength++;
        node = node.next;
    }

    const partLength = Math.floor(listLength / k);
    let leftLength = listLength % k;

    /** @type {(ListNode|null)[]} */
    const lists = [];

    const fakeHead = new ListNode(-1, head);

    for (let i = 0; i < k; i++) {
        node = fakeHead;
        for (let i = 0; i < partLength; i++) {
            node = node.next;
        }
        if (leftLength > 0) {
            node = node === null ? null : node.next;
            leftLength--;
        }

        if (node !== null) {
            const nextHead = node.next;
            node.next = null;
            lists.push(fakeHead.next);
            fakeHead.next = nextHead;
        } else {
            lists.push(null);
        }
    }

    return lists;
};
// @lc code=end

const _1 = new ListNode(1);
const _2 = new ListNode(2);
const _3 = new ListNode(3);

_1.next = _2;
_2.next = _3;

splitListToParts(_1, 2);

File: src/2023-2024/725.split-linked-list-in-parts.cpp
/*
 * @lc app=leetcode id=725 lang=cpp
 *
 * [725] Split Linked List in Parts
 */
#include <vector>

struct ListNode {
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next) {}
};

// @lc code=start
class Solution {
 public:
  std::vector<ListNode*> splitListToParts(ListNode* head, int k) {
    std::size_t listNodeLength = 0;
    ListNode* currentNode = head;
    while (currentNode != nullptr) {
      listNodeLength++;
      currentNode = currentNode->next;
    }

    const std::size_t partLength = listNodeLength / k;
    std::size_t remainder = listNodeLength % k;

    std::vector<ListNode*> parts;

    ListNode fakeHead(0, head);
    ListNode* prevNode = &fakeHead;
    currentNode = head;

    while (currentNode != nullptr) {
      std::size_t currentPartLength = partLength;
      if (remainder > 0) {
        currentPartLength++;
        remainder--;
      }

      parts.push_back(currentNode);

      for (auto i = 0; i < currentPartLength; i++) {
        prevNode = currentNode;
        currentNode = currentNode->next;
      }
      prevNode->next = nullptr;

      fakeHead.next = currentNode;
      prevNode = &fakeHead;
    }

    while (parts.size() < k) {
      parts.push_back(nullptr);
    }

    return parts;
  }
};
// @lc code=end
